docs/config.rst

.. _configuration:

=============
Configuration
=============

Many defaults can be configured via :py:obj:`csoundengine.config.config` (an instance 
of :class:`configdict.ConfigDict`, see https://configdict.readthedocs.io).

.. note::

    To modify defaults in a persistent manner, call :meth:`ConfigDict.save`
    to save the current version. The changes will be loaded the next time this
    module is imported

.. image:: assets/jupyter-config.png

Example
=======

.. code::

    import csoundengine as ce
    # Modify the default number of channels, force it to 2
    ce.config['nchnls'] = 2

    # Set default A4 value 
    ce.config['A4'] = 443

    # Larger number of audio buses
    ce.config['num_audio_buses'] = 200

    # If you would like to set these as defaults, call save. The next time using
    # csoundengine, these settings will be the default.
    ce.config.save()

    # To revert the changes to the default, use reset:
    ce.config.reset()

    
Any :class:`~csoundengine.engine.Engine` created after this will pick up these
new defaults::


    >>> from csoundengine import *
    >>> engine = Engine()
    >>> engine.numAudioBuses
    200
    >>> engine.A4
    443


Edit the config interactively
-----------------------------

.. code::

    from csoundengine import *
    config.edit()


This will open the config in a text editor and any changes there will be reflected back in
the config. Which application is used depends on the os settings for handling yaml files.

.. image:: assets/config-edit.png

-----

.. _configkeys:
    
Keys
====

.. include:: configkeys.rst


docs/configkeys.rst
.. _config_sr:

sr:
    | Default: **0**  -- ``int``
    | Choices: ``0, 22050, 24000, 44100, 48000, 88200, 96000, 144000, 192000``
    | *samplerate - 0=default sr for the backend*

.. _config_rec_sr:

rec_sr:
    | Default: **44100**  -- ``int``
    | Choices: ``0, 22050, 24000, 44100, 48000, 88200, 96000, 144000, 192000``
    | *default samplerate when rendering*

.. _config_nchnls:

nchnls:
    | Default: **0**  -- ``int``
    | Between 0 - 128
    | *Number of output channels. 0=default for used device*

.. _config_nchnls_i:

nchnls_i:
    | Default: **0**  -- ``int``
    | Between 0 - 128
    | *Number of input channels to use. 0 = default for used device*

.. _config_ksmps:

ksmps:
    | Default: **64**  -- ``int``
    | Choices: ``16, 32, 64, 128, 256, 512, 1024``
    | *corresponds to csound's ksmps*

.. _config_rec_ksmps:

rec_ksmps:
    | Default: **64**  -- ``int``
    | Choices: ``16, 32, 64, 128, 256, 512, 1024``
    | *samples per cycle when rendering*

.. _config_rec_sample_format:

rec_sample_format:
    | Default: **float**  -- ``(int, str)``
    | Choices: ``16, 24, 32, float``
    | *Sample format used when rendering*

.. _config_rec_suppress_output:

rec_suppress_output:
    | Default: **False**  -- ``bool``
    | *Supress debugging output when rendering offline*

.. _config_buffersize:

buffersize:
    | Default: **0**  -- ``int``
    | *-b value. 0=determine buffersize depending on ksmps & backend*

.. _config_numbuffers:

numbuffers:
    | Default: **0**  -- ``int``
    | *determines the -B value as a multiple of the buffersize. 0=auto*

.. _config_linux_backend:

linux_backend:
    | Default: **jack, pulse, pa_cb**  -- ``str``
    | *a comma separated list of backends (possible backends: jack, pulse, pa_cb, alsa)*

.. _config_macos_backend:

macos_backend:
    | Default: **pa_cb**  -- ``str``
    | *a comma separated list of backends (possible backends: pa_cb, auhal)*

.. _config_windows_backend:

windows_backend:
    | Default: **pa_cb**  -- ``str``
    | *a comma separated list of backends (possible backends: pa_cb, pa_bl)*

.. _config_a4:

A4:
    | Default: **442**  -- ``int``
    | Between 410 - 460
    | *Frequency for A4*

.. _config_check_pargs:

check_pargs:
    | Default: **False**  -- ``bool``
    | *Check number of pargs passed to instr*

.. _config_offline_score_table_size_limit:

offline_score_table_size_limit:
    | Default: **1900**  -- ``int``
    | *size limit when writing tables as f score statements via gen2. If a table is bigger than this size, it is saved as a datafile as gen23 or wav*

.. _config_fail_if_unmatched_pargs:

fail_if_unmatched_pargs:
    | Default: **False**  -- ``bool``
    | *Fail if the # of passed pargs doesnt match the # of pargs*

.. _config_set_sigint_handler:

set_sigint_handler:
    | Default: **True**  -- ``bool``
    | *Set a sigint handler to prevent csound crash with CTRL-C*

.. _config_generalmidi_soundfont:

generalmidi_soundfont:
    | Default: **''**  -- ``str``

.. _config_suppress_output:

suppress_output:
    | Default: **True**  -- ``bool``
    | *Suppress csound´s debugging information*

.. _config_unknown_parameter_fail_silently:

unknown_parameter_fail_silently:
    | Default: **True**  -- ``bool``
    | *Do not raise if a synth tries to set an unknown parameter*

.. _config_define_builtin_instrs:

define_builtin_instrs:
    | Default: **True**  -- ``bool``
    | *If True, a Session with have all builtin instruments defined*

.. _config_sample_fade_time:

sample_fade_time:
    | Default: **0.05**  -- ``float``
    | *Fade time when playing samples via a Session*

.. _config_prefer_udp:

prefer_udp:
    | Default: **True**  -- ``bool``
    | *If true and a server was defined prefer UDP over the API for communication*

.. _config_start_udp_server:

start_udp_server:
    | Default: **False**  -- ``bool``
    | *Start an engine with udp communication support*

.. _config_associated_table_min_size:

associated_table_min_size:
    | Default: **16**  -- ``int``
    | *Min. size of the param table associated with a synth*

.. _config_num_audio_buses:

num_audio_buses:
    | Default: **64**  -- ``int``
    | *Num. of audio buses in an Engine/Session*

.. _config_num_control_buses:

num_control_buses:
    | Default: **512**  -- ``int``
    | *Num. of control buses in an Engine/Session*

.. _config_html_theme:

html_theme:
    | Default: **light**  -- ``str``
    | Choices: ``dark, light``
    | *Style to use when displaying syntax highlighting*

.. _config_html_args_fontsize:

html_args_fontsize:
    | Default: **12px**  -- ``str``
    | *Font size used for args when outputing html (in jupyter)*

.. _config_synth_repr_max_args:

synth_repr_max_args:
    | Default: **12**  -- ``int``
    | *Max. number of pfields shown when in a synth's repr*

.. _config_synthgroup_repr_max_rows:

synthgroup_repr_max_rows:
    | Default: **4**  -- ``int``
    | *Max. number of rows for a SynthGroup repr. Use 0 to disable*

.. _config_jupyter_synth_repr_stopbutton:

jupyter_synth_repr_stopbutton:
    | Default: **True**  -- ``bool``
    | *When running inside a jupyter notebook, display a stop buttonfor Synths and SynthGroups*

.. _config_jupyter_synth_repr_interact:

jupyter_synth_repr_interact:
    | Default: **True**  -- ``bool``
    | *When inside jupyter, add interactive widgets if a synth hasnamed parameters*

.. _config_jupyter_instr_repr_show_code:

jupyter_instr_repr_show_code:
    | Default: **True**  -- ``bool``
    | *Show code when displaying an Instr inside jupyter*

.. _config_ipython_load_magics_at_startup:

ipython_load_magics_at_startup:
    | Default: **True**  -- ``bool``
    | *Load csoundengine.magic at startup when inside ipython. If False, magics can still be loaded via `%load_ext csoundengine.magic`*

.. _config_magics_print_info:

magics_print_info:
    | Default: **False**  -- ``bool``
    | *Print some informative information when the csounengine.magic extension is loaded*

.. _config_jupyter_slider_width:

jupyter_slider_width:
    | Default: **80%**  -- ``str``
    | *CSS Width used by an interactive slider in jupyter*

.. _config_timeout:

timeout:
    | Default: **2.0**  -- ``float``
    | *Timeout for any action waiting a response from csound*

.. _config_sched_latency:

sched_latency:
    | Default: **0.0**  -- ``float``
    | *Time delay added to any event scheduled to ensure that simultameous events arenot offset by scheduling overhead*

.. _config_datafile_format:

datafile_format:
    | Default: **gen23**  -- ``str``
    | Choices: ``gen23, wav``
    | *Format used when saving a table as a datafile*

.. _config_spectrogram_colormap:

spectrogram_colormap:
    | Default: **inferno**  -- ``str``
    | Choices: ``cividis, inferno, magma, plasma, viridis``

.. _config_samplesplot_figsize:

samplesplot_figsize:
    | Default: **12:4**  -- ``str``
    | *The figure size of the plot in the form '<width>:<height>'*

.. _config_spectrogram_figsize:

spectrogram_figsize:
    | Default: **24:8**  -- ``str``
    | *The figure size of the plot in the form '<width>:<height>'*

.. _config_spectrogram_maxfreq:

spectrogram_maxfreq:
    | Default: **12000**  -- ``int``
    | *Highest freq. in a spectrogram*

.. _config_spectrogram_window:

spectrogram_window:
    | Default: **hamming**  -- ``str``
    | Choices: ``hamming, hanning``

.. _config_dependencies_check_timeout_days:

dependencies_check_timeout_days:
    | Default: **7**  -- ``int``
    | Between 1 - 365
    | *Elapsed time (in days) after which dependencies will be checked*


csoundengine/config.py
import logging
from configdict import ConfigDict
from functools import lru_cache

modulename = 'csoundengine.engine'

logger = logging.getLogger('csoundengine')


def setLoggingLevel(level: str) -> None:
    """
    Utility to set the logging level of csoundengine
    """
    level = level.upper()
    logging.basicConfig(level=level)
    logger.setLevel(level)


# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
#                  CONFIG                   #
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #

_audioBackendsByPlatform = {
    'linux': ('jack', 'alsa', 'pa_cb', 'pa_bl', 'pulse'),
    'macos': ('auhal', 'pa_cb', 'pa_bl', 'jack'),
    'windows': ('pa_cb', 'pa_bl')
}


@lru_cache(maxsize=0)
def _validateBackend2(key:str, s: str) -> bool:
    platform = key.split("_")[0]
    possibleBackends = _audioBackendsByPlatform.get(platform)
    if not possibleBackends:
        logger.error(f"Platform {platform} not supported, no audio backends")
    for backend in (b.strip() for b in s.split(',')):
        if backend not in possibleBackends:
            logger.error(f"Backend '{backend}' unknown. "
                         f"Possible backends: {possibleBackends}")
            return False
    return True


def _validateBackend(cfg: dict, key:str, s: str) -> bool:
    return _validateBackend2(key, s)


def _validateFigsize(cfg: dict, key: str, val) -> bool:
    if not isinstance(val, str):
        return False
    parts = val.split(":")
    return len(parts) == 2 and all(p.isnumeric() for p in parts)

config = ConfigDict(modulename, persistent=False)
_ = config.addKey

_('sr', 0,
  choices=(0, 22050, 24000, 44100, 48000, 88200, 96000, 144000, 192000),
  doc='samplerate - 0=default sr for the backend')
_('rec_sr', 44100,
  choices=(0, 22050, 24000, 44100, 48000, 88200, 96000, 144000, 192000),
  doc='default samplerate when rendering')
_('nchnls', 0,
  range=(0, 128),
  doc='Number of output channels. 0=default for used device')
_('nchnls_i', 0,
  range=(0, 128),
  doc='Number of input channels to use. 0 = default for used device')
_('ksmps', 64,
  choices=(16, 32, 64, 128, 256, 512, 1024),
  doc="corresponds to csound's ksmps")
_('rec_ksmps', 64,
  choices=(16, 32, 64, 128, 256, 512, 1024),
  doc="samples per cycle when rendering")
_('rec_sample_format', 'float',
  choices=(16, 24, 32, 'float'),
  doc="Sample format used when rendering")
_('rec_suppress_output', False,
  doc='Supress debugging output when rendering offline')
_('buffersize', 0,
  doc="-b value. 0=determine buffersize depending on ksmps & backend")
_('numbuffers', 0,
  doc="determines the -B value as a multiple of the buffersize. 0=auto")
_('linux_backend', 'jack, pulse, pa_cb',
  doc="a comma separated list of backends (possible backends: jack, pulse, pa_cb, alsa)",
  validatefunc=_validateBackend),
_('macos_backend', 'pa_cb',
  doc="a comma separated list of backends (possible backends: pa_cb, auhal)",
  validatefunc=_validateBackend),
_('windows_backend', 'pa_cb',
  doc="a comma separated list of backends (possible backends: pa_cb, pa_bl)",
  validatefunc=_validateBackend)
_('A4', 442,
  range=(410, 460),
  doc="Frequency for A4")
_('check_pargs', False,
  doc='Check number of pargs passed to instr')
_('offline_score_table_size_limit', 1900,
  doc='size limit when writing tables as f score statements via gen2. If a table '
      'is bigger than this size, it is saved as a datafile as gen23 or wav')
_('fail_if_unmatched_pargs', False,
  doc='Fail if the # of passed pargs doesnt match the # of pargs'),
_('set_sigint_handler', True,
  doc='Set a sigint handler to prevent csound crash with CTRL-C')
_('generalmidi_soundfont', '')
_('suppress_output', True,
  doc='Suppress csound´s debugging information')
_('unknown_parameter_fail_silently', True,
  doc='Do not raise if a synth tries to set an unknown parameter')
_('define_builtin_instrs', True,
  doc="If True, a Session with have all builtin instruments defined")
_('sample_fade_time', 0.05,
  doc="Fade time when playing samples via a Session")
_("prefer_udp", True,
  doc="If true and a server was defined prefer UDP over the API for communication")
_("start_udp_server", False,
  doc="Start an engine with udp communication support")
_('associated_table_min_size', 16,
  doc="Min. size of the param table associated with a synth")
_('num_audio_buses', 64,
  doc="Num. of audio buses in an Engine/Session")
_('num_control_buses', 512,
  doc="Num. of control buses in an Engine/Session")
_('html_theme', 'light',
  choices={'dark', 'light'},
  doc="Style to use when displaying syntax highlighting")
_('html_args_fontsize', '12px',
  doc="Font size used for args when outputing html (in jupyter)")
_('synth_repr_max_args', 12,
  doc="Max. number of pfields shown when in a synth's repr")
_('synthgroup_repr_max_rows', 4,
  doc='Max. number of rows for a SynthGroup repr. Use 0 to disable')
_('jupyter_synth_repr_stopbutton', True,
  doc='When running inside a jupyter notebook, display a stop button'
      'for Synths and SynthGroups')
_('jupyter_synth_repr_interact', True,
  doc='When inside jupyter, add interactive widgets if a synth has'
      'named parameters')
_('jupyter_instr_repr_show_code', True,
  doc='Show code when displaying an Instr inside jupyter')
_('ipython_load_magics_at_startup', True,
  doc='Load csoundengine.magic at startup when inside ipython. If False, magics can '
      'still be loaded via `%load_ext csoundengine.magic`')
_('magics_print_info', False,
  doc='Print some informative information when the csounengine.magic extension is loaded')
_('jupyter_slider_width', '80%',
  doc='CSS Width used by an interactive slider in jupyter')
_('timeout', 2.,
  doc='Timeout for any action waiting a response from csound')
_('sched_latency', 0.,
  doc='Time delay added to any event scheduled to ensure that simultameous events are'
      'not offset by scheduling overhead')
_('datafile_format', 'gen23',
  choices={'gen23', 'wav'},
  doc='Format used when saving a table as a datafile')

# Plotting
_('spectrogram_colormap', 'inferno', choices={'viridis', 'plasma', 'inferno', 'magma', 'cividis'})
_('samplesplot_figsize', "12:4", validatefunc=_validateFigsize,
  doc="The figure size of the plot in the form '<width>:<height>'")
_('spectrogram_figsize', "24:8", validatefunc=_validateFigsize,
  doc="The figure size of the plot in the form '<width>:<height>'")
_('spectrogram_maxfreq', 12000,
  doc="Highest freq. in a spectrogram")
_('spectrogram_window', 'hamming', choices={'hamming', 'hanning'})
_('dependencies_check_timeout_days', 7,
  doc="Elapsed time (in days) after which dependencies will be checked",
  range=(1, 365))

config.load()

